lib/pbench/server/database/alembic.ini

```diff
@@ -35,7 +35,7 @@ script_location = alembic
 # are written from script.py.mako
 # output_encoding = utf-8
 
-sqlalchemy.url = driver://user:pass@localhost/dbname
+sqlalchemy.url = postgresql://pbench:pbench@localhost:5432/pbench
```

Comment:
An interesting choice: where is `localhost` going to be a useful default? Even in our CI pod deployment it's on a different host in the virtual pod network, and this is presumably running in the Pbench Server container.

Another choice might be to use a trivial placeholder and *sed* it from the config file during container deployment.

Reply:
Yeah, sure; OK, the point is to "use the Source, Luke" at CI time to build a PostgreSQL schema in a local "scratch" DB so we can save the Alembic file.
1. I still don't like duplicating a separate `pbench-server.cfg` file: I don't think that's a good idea. (And many of the settings here are nonsensical in that environment.)
2. There's still no point if we don't have a mechanism to save the resulting Alembic file into the container when we build it, so that the container can start up and use it to build/update the schema.

Reply:
I am happy to drop this change entirely.  I can just put in the commit message how I changed the files to make this work.

Reply:
No, this may be iterating towards a useful first step.

Maybe, instead of leaving things undefined and commented out as a placeholder, implement `alembic.run` to generate a version file based on the current sources. Make a `tox` target that uses it to generate an update file.

We still need to add a process to get that into the container, and to run it during startup against the real configured DB; and ideally the CI will be generating them automatically so they can be verified during testing and automatically deployed by the CD scripting. The Flask migration stuff may be the way we get it executed during startup; but having the file is a first step.

I still don't like the config file copy, here, but maybe that's a problem with our DB init logic that's designed to pull the PSQL URL out of a config file rather than simply being given the URL. If we fixed *that* (here or in a separate PR), then Alembic wouldn't need a config file at all.

Reply:
Okay, so I'll update the alembic.run so that if somebody uses `tox -e alembic` it will generate a new revision against the configured PostgreSQL database.  Then, the developer can check the file contents to see if it did anything worth committing.

And, yes, I'll be posting a follow-on commit which will solve that problem on top of this foundational capture of the "empty -> current" database migration.

Reply:
Yeah, I agree with Dave here. Also if you are hard-coding the `sqlalchemy.url` why do we need a copy of pbench-server.cfg file in the alembic directory?

Reply:
I see at least three possible "styles" for managing this, going forward. One basic agile rule, for example, is that at any time the top of `main` should be "ship-able"; and with a full CD pipeline we might actually do that.
1. Automate it under CD so that each *deployment* generates an update alembic file from the previous deployment, puts it back in the repo, and deploys it;
2. Automate it under the CI so that each merge onto main which changes the DB schema persists an update file;
3. Require that each PR that changes the schema includes an alembic update. We can enforce this by running the alembic script in the CI and *failing* if it's not "empty".

I'm happy with all of that being "future", but having your `tox` script actually generate a "top of branch" alembic file based on the latest in the repo sounds like a good first step for any direction.

Reply:
I don't want to rain on anybody's parade here or keep a "first step" from being _taken_, but I have to ask:  when we do a migration, what are we migrating **_from_**?

1. That is, it's all well and good, for the CD to put something into the repo, but, once the CD's deployment happens, the thing added to the repo will now be passe, and it will need to be _removed_ again...right?
2. Likewise, having the CI persist schema changes sounds like a great idea, but...are those changes cumulative?  If so, from where?  And, what's the mechanism for dropping those changes?  (I.e., after we have one or two, won't subsequent deployments be broken with migration failures?)
3. I like the notion that the CI will detect changes which will result in DB migrations...but, again, I'm not sure what it should _do_ with them.

I feel like the CI should be setting stuff up for the CD handler to dispose of...which somehow seems doable to me.  But, it makes the notion of running continuous deployment foundational, which is going to be tricky vis a vis running both a staging server and a production server.  (I.e., presumably the CD will maintain the staging server, which means we'll _also_ need a mechanism for migrating the production server when we re-deploy that...unless we're going to buy into full-on CD (i.e., no staging, just production)...which, from this angle, actually looks attractive!)

Reply:
The alembic update files are cumulative; the only real issue is how granular we want to be. E.g., in a waterfall release-based product, one might generate update files only at a release, which (via the cumulative logic) would create green field DBs, or update any previous release.

In an agile product, especially with CD, we need to be careful about this. In order to *update* a managed service we need the currently deployed schema as well as the new desired schema. Alembic can do this automatically from the live DB and the sources, or we can make sure we have a checkpoint matching what was deployed last (e.g., the update file history in the repo).

The hard part is to decide what we *want* to do.

Reply:
> The alembic update files are cumulative; the only real issue is how granular we want to be. E.g., in a waterfall release-based product, one might generate update files only at a release, which (via the cumulative logic) would create green field DBs, or update any previous release.

I don't think I understand:  if an update file calls for a change which is already present in the database, does Alembic ignore the change or flag it as an error?

That is, if we come out with v1.8 after changing the DB in v1.7 and v1.6, and we apply the cumulative updates since v1.0 to a DB which is already updated to v1.7, will that produce a v1.8 DB or a failure??

> In an agile product, especially with CD, we need to be careful about this. In order to _update_ a managed service we need the currently deployed schema as well as the new desired schema. Alembic can do this automatically from the live DB and the sources, or we can make sure we have a checkpoint matching what was deployed last (e.g., the update file history in the repo).

That checkpoint is what I was getting at with my comments about maintaining both a Staging server and a Production server.  Being able to express what we want the schema to be and having Alembic apply that to an existing DB without us having to describe the transition would be great.  Otherwise, as you say, we (meaning the CI and/or CD handler) need to know what the migration is coming _from_, so that we can request that Alembic make only the necessary updates...I think....

Reply:
See if this latest round of changes is work-a-bull.

I am using the latest version of alembic, `1.9.x`, which provides the pretty cool command, `alembic check`, which I integrated into our existing `tox` environment via `exec-tests`.  That way, the `tox` check will fail if a commit contains un-processed migrations.